File: demo-ios-objc/demo-ios-objc/GAMBannerViewController.m
//
//  GAMBannerViewController.m
//  demo-ios-objc
//
//  Copyright © 2020 Optable Technologies Inc. All rights reserved.
//  See LICENSE for details.
//

#import "OptableSDKDelegate.h"
#import "GAMBannerViewController.h"
#import "AppDelegate.h"
#import "demo_ios_objc-Swift.h"

@import OptableSDK;
@import GoogleMobileAds;

@interface GAMBannerViewController ()
// GoogleMobileAds - GADBannerView
@property(nonatomic, strong) GADBannerView *gadBannerView;
// Logging
@property(nonatomic, strong, nullable) NSString* targetingLog;
@property(nonatomic, strong, nullable) NSString* witnessLog;
@property(nonatomic, strong, nullable) NSString* profileLog;
@property(nonatomic, strong, nullable) id networkLogObserver;
@end

@implementation GAMBannerViewController

- (NSString *)AD_MANAGER_AD_UNIT_ID {
    return @"/22081946781/ios-sdk-demo/mobile-leaderboard";
}

- (void)viewDidLoad {
    [super viewDidLoad];

    self.gadBannerView = [[GADBannerView alloc] initWithAdSize:GADAdSizeBanner];
    self.gadBannerView.adUnitID = self.AD_MANAGER_AD_UNIT_ID;
    self.gadBannerView.rootViewController = self;
    self.gadBannerView.delegate = self;
    [self addBannerViewToView:self.gadBannerView];

    OptableSDKDelegate *delegate = (OptableSDKDelegate *)OPTABLE.delegate;
    delegate.gadBannerView = self.gadBannerView;
    delegate.pbmBannerAdUnit = nil;
    delegate.targetingOutput = self.targetingOutput;
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear: animated];
    [self startObservingNetworkLogs];
}

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear: animated];
    [self stopObservingNetworkLogs];
}

// MARK: - Actions
- (IBAction)loadBannerWithTargeting:(id)sender {
    NSError *error = nil;
    [OPTABLE targetingWithIds: NULL error: &error];
    [OPTABLE witnessWithEvent: @"GAMBannerViewController.loadBannerClicked"
                   properties: @{ @"example": @"value" }
                        error: &error];
    [OPTABLE profileWithTraits: @{ @"example": @"value", @"anotherExample": @123, @"thirdExample": @YES }
                            id: NULL
                     neighbors: NULL
                         error: &error];
}

- (IBAction)loadBannerWithTargetingFromCache:(id)sender {
    NSError *error = nil;
    GAMRequest *request = [GAMRequest request];
    NSDictionary *cachedTargetingData = [OPTABLE targetingFromCache];

    if (cachedTargetingData != nil) {
        request.customTargeting = cachedTargetingData;
        NSLog(@"[OptableSDK] ✅ Cached targeting values found: %@", cachedTargetingData);
    } else {
        NSLog(@"[OptableSDK] ℹ️ Cache empty");
    }
    
    [(OptableSDKDelegate*)OPTABLE.delegate loadGADAdWithTargetingData:cachedTargetingData];
    [OPTABLE witnessWithEvent: @"GAMBannerViewController.loadBannerClicked"
                   properties: @{ @"example": @"value" }
                        error: &error];
    [OPTABLE profileWithTraits: @{ @"example": @"value", @"anotherExample": @123, @"thirdExample": @YES }
                            id: NULL
                     neighbors: NULL
                         error: &error];
}

- (IBAction)clearTargetingCache:(id)sender {
    [_targetingOutput setText:@"🧹 Cleared local targeting cache.\n"];
    [OPTABLE targetingClearCache];
}

// MARK: - GADBannerViewDelegate
- (void)bannerView:(GADBannerView *)bannerView didFailToReceiveAdWithError:(NSError *)error {
    NSLog(@"[GAMBannerViewController] Failed to receive ad: %@", [error localizedDescription]);
}

// MARK: - Helpers
- (void)addBannerViewToView:(UIView *)bannerView {
    bannerView.translatesAutoresizingMaskIntoConstraints = NO;
    [self.adPlaceholder addSubview:bannerView];
    
    [NSLayoutConstraint activateConstraints:@[
        [bannerView.centerXAnchor constraintEqualToAnchor:self.adPlaceholder.centerXAnchor],
        [bannerView.centerYAnchor constraintEqualToAnchor:self.adPlaceholder.centerYAnchor]
    ]];
}

- (void)updateUILog {
    NSMutableArray<NSString *> *parts = [NSMutableArray array];

    if (self.targetingLog) {
        [parts addObject:self.targetingLog];
    }
    if (self.witnessLog) {
        [parts addObject:self.witnessLog];
    }
    if (self.profileLog) {
        [parts addObject:self.profileLog];
    }

    self.targetingOutput.text = [parts componentsJoinedByString:@"\n\n"];
}

// MARK: - Logging
- (void)startObservingNetworkLogs {
    __weak typeof(self) weakSelf = self;
    
    _networkLogObserver = [NSNotificationCenter.defaultCenter
                           addObserverForName: NotificationNames.HTTPURLLogUpdated
                           object: nil
                           queue: [NSOperationQueue mainQueue]
                           usingBlock: ^(NSNotification* notification) {
        HTTPURLLogEntry *logEntry = notification.userInfo[@"data"];
        if (![logEntry isKindOfClass:[HTTPURLLogEntry class]]) {
            return;
        }
        
        NSString *urlString = logEntry.request.URL.absoluteString;
        if ([urlString containsString:@"/targeting"]) {
            weakSelf.targetingLog = logEntry.debugDescription;
            if (logEntry.response == nil) {
                NSLog(@"%@", logEntry.requestDebugDescription);
            } else {
                NSLog(@"%@", logEntry.responseDebugDescription);
            }
        }
        if ([urlString containsString:@"/witness"]) {
            weakSelf.witnessLog = logEntry.debugDescription;
            if (logEntry.response == nil) {
                NSLog(@"%@", logEntry.requestDebugDescription);
            } else {
                NSLog(@"%@", logEntry.responseDebugDescription);
            }
        }
        if ([urlString containsString:@"/profile"]) {
            weakSelf.profileLog = logEntry.debugDescription;
            if (logEntry.response == nil) {
                NSLog(@"%@", logEntry.requestDebugDescription);
            } else {
                NSLog(@"%@", logEntry.responseDebugDescription);
            }
        }
        
        [weakSelf updateUILog];
    }];
}

- (void)stopObservingNetworkLogs {
    if (_networkLogObserver != NULL) {
        [NSNotificationCenter.defaultCenter removeObserver: _networkLogObserver];
    }
}

@end

File: demo-ios-swift/demo-ios-swift/GAMBannerViewController.swift
//
//  GAMBannerViewController.swift
//  demo-ios-swift
//
//  Copyright © 2020 Optable Technologies Inc. All rights reserved.
//  See LICENSE for details.
//

import GoogleMobileAds
import OptableSDK
import UIKit

private let AD_MANAGER_AD_UNIT_ID = "/22081946781/ios-sdk-demo/mobile-leaderboard"

// MARK: - GAMBannerViewController
final class GAMBannerViewController: UIViewController {
    // Outlets
    @IBOutlet var adPlaceholder: UIView!
    @IBOutlet var loadBannerButton: UIButton!
    @IBOutlet var loadBannerFromCacheButton: UIButton!
    @IBOutlet var clearTargetingCacheButton: UIButton!
    @IBOutlet var targetingOutput: UITextView!

    // GoogleMobileAds - GADBannerView
    var gadBannerView: BannerView!

    // Logging
    private var targetingLog: String? { didSet { updateUILog() } }
    private var witnessLog: String? { didSet { updateUILog() } }
    private var profileLog: String? { didSet { updateUILog() } }
    private var networkLogObserver: (any NSObjectProtocol)?

    override func viewDidLoad() {
        super.viewDidLoad()

        gadBannerView = BannerView(adSize: AdSizeBanner)
        gadBannerView.adUnitID = AD_MANAGER_AD_UNIT_ID
        gadBannerView.rootViewController = self
        gadBannerView.delegate = self
        addBannerViewToView(gadBannerView)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        startObservingNetworkLogs()
    }

    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        stopObservingNetworkLogs()
    }

    // MARK: Actions
    @IBAction func loadBannerWithTargeting(_ sender: UIButton) {
        do {
            try OPTABLE!.targeting { [weak self] result in
                switch result {
                case let .success(optableTargeting):
                    print("[OptableSDK] ✅ Success on /targeting API call: \(optableTargeting)")
                    self?.loadBanner(optableTargeting)

                case let .failure(error):
                    print("[OptableSDK] 🚫 Error on /targeting API call: \(error)")
                    self?.loadBanner()
                }
            }
        } catch {
            print("[OptableSDK] 🚫 Exception on /targeting API call: \(error)")
            targetingLog = "🚫 EXCEPTION: \(error)"
        }
    }

    @IBAction func loadBannerWithTargetingFromCache(_ sender: UIButton) {
        if let cachedOptableTargeting = OPTABLE!.targetingFromCache() {
            print("[OptableSDK] ✅ Cached targeting values found: \(cachedOptableTargeting)")
            loadBanner(cachedOptableTargeting)
        } else {
            print("[OptableSDK] ℹ️ Cache empty")
            loadBanner()
        }
    }

    @IBAction func clearTargetingCache(_ sender: UIButton) {
        targetingOutput.text = "🧹 Cleared local targeting cache.\n"
        OPTABLE!.targetingClearCache()
    }
}

// MARK: - Private
private extension GAMBannerViewController {
    func loadBanner(_ optableTargeting: OptableTargeting? = nil) {
        loadGADAd(optableTargeting)
        witness()
        profile()
    }

    func loadGADAd(_ optableTargeting: OptableTargeting? = nil) {
        let adRequest = AdManagerRequest()
        adRequest.customTargeting = optableTargeting?.gamTargetingKeywords as? [String: Any]
        gadBannerView.load(adRequest)
    }

    func witness() {
        do {
            try OPTABLE!.witness(
                event: "GAMBannerViewController.loadBannerClicked",
                properties: ["example": "value"]
            ) { result in
                switch result {
                case .success:
                    print("[OptableSDK] ✅ Success on /witness API call")
                case let .failure(error):
                    print("[OptableSDK] 🚫 Error on /witness API call: \(error)")
                }
            }
        } catch {
            print("[OptableSDK] 🚫 Exception on /witness API call: \(error)")
            witnessLog = "🚫 EXCEPTION: \(error)"
        }
    }

    func profile() {
        do {
            try OPTABLE!.profile(
                traits: ["example": "value", "anotherExample": 123, "thirdExample": true]
            ) { result in
                switch result {
                case .success:
                    print("[OptableSDK] ✅ Success on /profile API call")
                case let .failure(error):
                    print("[OptableSDK] 🚫 Error on /profile API call: \(error)")
                }
            }
        } catch {
            print("[OptableSDK] 🚫 Exception on /profile API call: \(error)")
            profileLog = "🚫 EXCEPTION: \(error)"
        }
    }
}

// MARK: - GoogleMobileAds.BannerViewDelegate
extension GAMBannerViewController: GoogleMobileAds.BannerViewDelegate {
    func bannerView(_ bannerView: BannerView, didFailToReceiveAdWithError error: any Error) {
        print("[GAMBannerViewController] Failed to receive ad: \(error)")
    }
}

// MARK: - Helpers
private extension GAMBannerViewController {
    func addBannerViewToView(_ bannerView: BannerView) {
        bannerView.translatesAutoresizingMaskIntoConstraints = false
        adPlaceholder.addSubview(bannerView)

        NSLayoutConstraint.activate([
            bannerView.centerXAnchor.constraint(equalTo: adPlaceholder.centerXAnchor),
            bannerView.centerYAnchor.constraint(equalTo: adPlaceholder.centerYAnchor),
        ])
    }

    func updateUILog() {
        targetingOutput.text = [targetingLog, witnessLog, profileLog].compactMap({ $0 }).joined(separator: "\n\n")
    }
}

// MARK: - Logging
private extension GAMBannerViewController {
    /// Setups observation for URLRequest/URLResponse logging
    func startObservingNetworkLogs() {
        networkLogObserver = NotificationCenter.default
            .addObserver(forName: .HTTPURLLogUpdated, object: nil, queue: .main, using: { [weak self] notification in
                if let logEntry = notification.userInfo?["data"] as? HTTPURLLogEntry,
                   logEntry.request.url?.absoluteString.contains("/targeting") == true {
                    self?.targetingLog = logEntry.debugDescription
                    print(logEntry.response == nil ? logEntry.requestDebugDescription : logEntry.responseDebugDescription)
                }

                if let logEntry = notification.userInfo?["data"] as? HTTPURLLogEntry,
                   logEntry.request.url?.absoluteString.contains("/witness") == true {
                    self?.witnessLog = logEntry.debugDescription
                    print(logEntry.response == nil ? logEntry.requestDebugDescription : logEntry.responseDebugDescription)
                }

                if let logEntry = notification.userInfo?["data"] as? HTTPURLLogEntry,
                   logEntry.request.url?.absoluteString.contains("/profile") == true {
                    self?.profileLog = logEntry.debugDescription
                    print(logEntry.response == nil ? logEntry.requestDebugDescription : logEntry.responseDebugDescription)
                }
            })
    }

    func stopObservingNetworkLogs() {
        guard let networkLogObserver else { return }
        NotificationCenter.default.removeObserver(networkLogObserver)
    }
}
